java/ql/lib/semmle/code/java/frameworks/android/OnActivityResultSource.qll

```diff
@@ -0,0 +1,108 @@
+/** Provides a remote flow source for Android's `Activity.onActivityResult` method. */
+
+import java
+private import semmle.code.java.dataflow.DataFlow
+private import semmle.code.java.dataflow.DataFlow5
+private import semmle.code.java.frameworks.android.Android
+private import semmle.code.java.frameworks.android.Fragment
+private import semmle.code.java.frameworks.android.Intent
+
+/**
+ * The data Intent parameter in the `onActivityResult` method.
+ */
+class OnActivityResultIncomingIntent extends DataFlow::Node {
+  OnActivityResultIncomingIntent() {
+    exists(Method onActivityResult |
+      onActivityResult.getDeclaringType() instanceof ActivityOrFragment and
+      onActivityResult.hasName("onActivityResult") and
+      this.asParameter() = onActivityResult.getParameter(2)
+    )
+  }
+
+  /**
+   * Holds if this node is a remote flow source.
+   *
+   * This is only a source when the Activity or Fragment that implements `onActivityResult` is
+   * also using an implicit Intent to start another Activity with `startActivityForResult`. This
+   * means that a malicious application can intercept it to start itself and return an arbitrary
+   * Intent to `onActivityResult`.
+   */
+  predicate isRemoteSource() {
+    exists(ImplicitStartActivityForResultConf conf, RefType startingType, DataFlow::Node sink |
+      conf.hasFlowTo(sink) and
+      startingType = sink.asExpr().(Argument).getCall().getEnclosingCallable().getDeclaringType()
+    |
+      startingType = this.getEnclosingCallable().getDeclaringType()
+      or
+      // A fragment calls `startActivityForResult`
+      // and the activity it belongs to defines `onActivityResult`.
+      exists(MethodAccess ma |
+        ma.getMethod().hasName(["add", "attach", "replace"]) and
+        ma.getMethod().getDeclaringType().hasName("FragmentTransaction") and
+        any(Argument arg | arg = ma.getAnArgument()).getType() = startingType
+        or
+        ma.getMethod().hasName("show") and
+        ma.getMethod().getDeclaringType().getASupertype*().hasName("DialogFragment") and
+        startingType = ma.getQualifier().getType()
+      |
+        ma.getEnclosingCallable().getDeclaringType() =
+          this.getEnclosingCallable().getDeclaringType()
+      )
+    )
+  }
+}
+
+/**
+ * A data flow configuration for implicit intents being used in `startActivityForResult`.
+ */
+private class ImplicitStartActivityForResultConf extends DataFlow5::Configuration {
```

Comment:
Has a dataflow config --> file should have `Query.qll` name suffix (there are CI checks to avoid globally importing these)

Reply:
The intention was for this to be a globally applicable remote source, but this configuration stops us from doing that. Would you suggest renaming this to `*Query.qll`, explicitly importing it in all Android queries and making sure those don't define a `DataFlow5::Configuration`?

Reply:
Ah if it really should be global then it's fine, leave as is. It's good to use a dataflow config instance that no queries use, to avoid the results being needlessly recomputed. In one case I made an instance specially for that purpose (https://github.com/github/codeql/blob/main/java/ql/lib/semmle/code/java/dataflow/internal/DataFlowForSerializability.qll) to avoid unnecessary overlap, anticipating that getting deleted one we have good enough parameterised modules that a fresh instance is created cleanly and easily.

Reply:
I don't think there's any `DataFlowX::Configuration` not being currently used by at least one other query. Would you recommend adding another instance of `DataFlow` just for this specific case?

Reply:
Personally I think that's the best thing to do, though there may be counter-argument. Suggest raise it in the codeql channel before doing the work.

Reply:
Done in 8a4de3f.

Reply:
Don't forget to update https://github.com/github/codeql/blob/main/config/identical-files.json

Reply:
Oops, done in 6da062b. Thanks for keeping an eye on this.